src/state.rs

```diff
@@ -29,69 +30,81 @@ pub struct State {
-        let instance = wgpu::Instance::new(wgpu::BackendBit::PRIMARY);
+    ) -> (
+        wgpu::SurfaceConfiguration,
+        wgpu::Surface,
+        wgpu::Adapter,
+        wgpu::Device,
+        wgpu::Queue,
+    ) {
+        let instance = wgpu::Instance::new(wgpu::Backends::PRIMARY);
         let render_surface = unsafe { instance.create_surface(window) };
+
         let adapter = instance
             .request_adapter(&wgpu::RequestAdapterOptions {
-                power_preference: wgpu::PowerPreference::HighPerformance,
+                power_preference: wgpu::PowerPreference::default(),
                 compatible_surface: Some(&render_surface),
+                force_fallback_adapter: false,
             })
             .await
-            .unwrap();
-        println!("Using {:?}", adapter.get_info().backend);
+            .or_else(|| {
+                let adapters = instance.enumerate_adapters(wgpu::Backends::all());
+                eprintln!(
+                    "No matching graphics adapter available, using any: {:?}",
+                    adapters.collect::<Vec<_>>()
+                );
+                let mut adapters = instance.enumerate_adapters(wgpu::Backends::all());
+                adapters.next()
+            })
+            .expect("No graphics adapter");
+
+        println!(
+            "Using backend {:?} with features {:?}",
+            adapter.get_info().backend,
+            adapter.features()
+        );
 
         let (render_device, queue) = adapter
             .request_device(
                 &wgpu::DeviceDescriptor {
                     label: Some("render_device"),
-                    features: wgpu::Features::SAMPLED_TEXTURE_BINDING_ARRAY,
+                    features: wgpu::Features::TEXTURE_BINDING_ARRAY,
                     limits: wgpu::Limits::default(),
                 },
                 None,
             )
             .await
             .unwrap();
 
-        (render_surface, adapter, render_device, queue)
-    }
-
-    fn create_swap_chain(
-        window: &Window,
-        adapter: &wgpu::Adapter,
-        render_device: &wgpu::Device,
-        render_surface: &wgpu::Surface,
-    ) -> (wgpu::SwapChainDescriptor, wgpu::SwapChain) {
         let size = window.inner_size();
 
-        let swap_chain_descriptor = wgpu::SwapChainDescriptor {
-            usage: wgpu::TextureUsage::RENDER_ATTACHMENT,
-            format: adapter
-                .get_swap_chain_preferred_format(render_surface)
-                .unwrap(),
+        let config = wgpu::SurfaceConfiguration {
+            usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
+            format: render_surface.get_preferred_format(&adapter).unwrap(),
             width: size.width,
             height: size.height,
-            present_mode: wgpu::PresentMode::Immediate,
+
+            // was Immediate, but can't be guaranteed to work
+            present_mode: wgpu::PresentMode::Fifo,
```

Comment:
Did this break on your device? I'm using `Immediate` in testing because `Fifo` caps the framerate at vsync, which makes it hard to see what impact your changes have on performance, and according to the docs:

>`Immediate`
>
>The presentation engine does not wait for a vertical blanking period and the request is presented immediately. This is a low-latency presentation mode, but visible tearing may be observed. **Will fallback to `Fifo` if unavailable on the selected platform and backend.** Not optimal for mobile."

Reply:
Everything broke on my system and this was one of the things I tried to fix it, but it did nothing so I'll revert it. Good catch